### DISCUSSION TITLE
root user: copy to clipboard, not pasteable outside Turbo

### DISCUSSION THREAD

**Discussion by grahamperrin** [magiblot/turbo#70] [Q&A]
Using [editors/turbo](https://www.freshports.org/editors/turbo/), the port to FreeBSD, in Konsole. 

If I run `turbo` as myself: no problem. 

If I `su -` (become the superuser) before a run: 

- copy and paste work as expected within Turbo
- what I copy to the clipboard within Turbo is **not** on the clipboard outside Turbo. 

```
% uname -aKU ; pkg iinfo kde5 konsole ^turbo$
FreeBSD mowa219-gjp4-zbook-freebsd 15.0-CURRENT FreeBSD 15.0-CURRENT main-n268104-d04abb05375d GENERIC-NODEBUG amd64 1500013 1500013
kde5-5.27.10.23.08.4_2
konsole-23.08.4_1
turbo-g20230621
% 
```

Re: runtime requirements <https://github.com/magiblot/tvision?tab=readme-ov-file#build-linux-runtime>, 

```text
% pkg iinfo xclip xsel
xclip-0.13.82_1
xsel-conrad-1.2.0_2
% pkg rquery %o xsel-conrad
x11/xsel-conrad
% 
```

<https://www.freshports.org/x11/xclip/#message> no package message (I assume, nothing required after installation). 

[x11/xsel-conrad](https://www.freshports.org/x11/xsel-conrad/)

> Access X selection from command line

<https://www.freshports.org/x11/xsel-conrad/#message> no package message. 

----

I see these, but can't tell whether they're relevant: 

- https://github.com/magiblot/turbo/issues/2
- <https://www.scintilla.org/ScintillaDoc.html#CutCopyAndPaste>
- …

**Comment by magiblot** [comment]
Hi Graham!

In X11 desktop environments, clipboard interaction is done through `xsel` or `xclip`, which are X11 clients. For them to be able to connect to the X11 server, the `DISPLAY` environment variable must be properly defined.

When you run a terminal, the command-line shell properly inherits the `DISPLAY` variable. This also works when using `sudo`. However, the `su` command with the `-` option clears environment variables, as explained in its manpage:

```
       -, -l, --login
           Start the shell as a login shell with an environment similar to a real login:

           •   clears all the environment variables except TERM and variables specified by --whitelist-environment
```

Which results in the `DISPLAY` variable not being set inside the superuser subshell.

Therefore, you have to either drop the `-` option (if you don't need it) or add the `-w DISPLAY` option when running `su`.

Cheers.

**Reply by grahamperrin** [reply]
Thanks, now I see that Turbo does work as expected with sudo. 

----

> … or add the `-w DISPLAY` option when running `su`.

No `-w` in su(1) at <https://man.freebsd.org/cgi/man.cgi?query=su&sektion=1&manpath=freebsd-release>, am I missing something? 

If `su -` is followed by `DISPLAY` set to `:0`, should that work?

**Reply by magiblot** [reply]
Oops, you are right, my bad. The `-w` option is GNU-specific and it isn't available on FreeBSD.

Yes, setting `DISPLAY` to `:0` manually will work in most cases, since `DISPLAY` usually gets that value. But one example of a situation where this wouldn't work is when using X11 forwarding via SSH, since `DISPLAY` may get a different value in the remote shell session.

So a generic and portable way to run `su -` while preserving the `DISPLAY` variable could be:

```sh
su - -c "export DISPLAY=$DISPLAY; \$SHELL"
```

**Reply by magiblot** [reply]
Additionally, if your system disallows graphical applications being run as root, as described in https://stackoverflow.com/questions/48833451/no-protocol-specified-when-running-a-sudo-su-app-on-ubuntu-linux, you will have to run the following before `su`:

```sh
xhost si:localuser:root
```

Cheers.
